server/base/src/main/java/org/apache/accumulo/server/metadata/iterators/TabletMetadataCheckIterator.java

```diff
@@ -0,0 +1,191 @@
+import org.apache.accumulo.core.client.IteratorSetting;
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Condition;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.dataImpl.KeyExtent;
+import org.apache.accumulo.core.iterators.IteratorAdapter;
+import org.apache.accumulo.core.iterators.IteratorEnvironment;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.accumulo.core.metadata.schema.TabletMetadata;
+import org.apache.accumulo.core.metadata.schema.TabletMetadataCheck;
+import org.apache.accumulo.server.metadata.ConditionalTabletMutatorImpl;
+import org.apache.hadoop.io.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+public class TabletMetadataCheckIterator implements SortedKeyValueIterator<Key,Value> {
+
+  private static final Logger log = LoggerFactory.getLogger(TabletMetadataCheckIterator.class);
+
+  private TabletMetadataCheck check;
+  private KeyExtent expectedExtent;
+
+  private Key startKey;
+  private Value topValue;
+  private SortedKeyValueIterator<Key,Value> source;
+
+  private static final String CHECK_CLASS_KEY = "checkClass";
+  private static final String CHECK_DATA_KEY = "checkData";
+  private static final String EXTENT_KEY = "checkExtent";
+  private static final String SUCCESS = "success";
+
+  @Override
+  public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options,
+      IteratorEnvironment env) throws IOException {
+    Preconditions.checkState(check == null && expectedExtent == null && startKey == null);
+    try {
+      String className = Objects.requireNonNull(options.get(CHECK_CLASS_KEY));
+      String checkData = Objects.requireNonNull(options.get(CHECK_DATA_KEY));
+      log.trace("Instantiating class {} using {}", className, checkData);
+      Class<? extends TabletMetadataCheck> clazz =
+          ClassLoaderUtil.loadClass(null, className, TabletMetadataCheck.class);
+      check = GSON.get().fromJson(checkData, clazz);
+      expectedExtent = KeyExtent.fromBase64(options.get(EXTENT_KEY));
+      this.source = source;
+    } catch (ReflectiveOperationException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Override
+  public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive)
+      throws IOException {
+    Preconditions.checkState(check != null && expectedExtent != null);
+
+    this.startKey = null;
+
+    Text tabletRow = getTabletRow(range);
+
+    var expectedMetaRow = expectedExtent.toMetaRow();
+    Preconditions.checkState(tabletRow.equals(expectedMetaRow), "Tablet row mismatch %s %s",
+        tabletRow, expectedMetaRow);
+
+    var colsToRead = check.columnsToRead();
+    // TODO actually use columns in seek and TabletMetadata construction
```

Comment:
Need to remove this TODO and open a follow on issue before merging

Reply:
I was actually just getting ready to ask you if you planned to do this now or in a future PR

Reply:
Opened #5254 and removed the TODO